src/source/channel_router.rs

```diff
@@ -0,0 +1,306 @@
+// Channel router types and implementation.
+
+use crate::{Sample, Source};
+use std::{
+    cmp::min,
+    error::Error,
+    fmt,
+    sync::mpsc::{channel, Receiver, Sender},
+};
+
+/// A matrix to map inputs to outputs according to a gain
+///
+/// A two-dimensional matrix of `f32`s:
+/// - The first dimension is respective to the input channels
+/// - The second is respective to the output channels
+///
+/// Thus, if a value at `map[1,1]` is 0.2, this signifies that the signal on
+/// channel 1 should be mixed into channel 1 with a coefficient of 0.2.
+pub type ChannelMap = Vec<Vec<f32>>;
+// doing this as Vec<Vec<atomic_float::AtomicF32>> would require feature=experimental, so I decided
+// to just use a channel to do updates.
+//
+// Doing it as a HashMap<(u16,u16), f32> is an option too but there's a penalty hashing these
+// values, there's ways to speed that up though. It'd be great if the object upgraded its
+// implementation if it got sufficiently big.
```

Comment:
No we do not need complex containers, if a sparse matrix is preferable I have an [example](https://github.com/RustAudio/rodio/issues/653#issuecomment-2526344738) how that can be implemented.